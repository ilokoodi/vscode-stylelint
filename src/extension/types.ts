// eslint-disable-next-line node/no-unpublished-import
import type TypedEventEmitter from 'typed-emitter';

/**
 * VS Code extension event names.
 */
export interface ExtensionEvents {
	DidRegisterDocumentFormattingEditProvider: () => void;
}

/**
 * VS Code extension public API.
 */
export type PublicApi = TypedEventEmitter<ExtensionEvents> & {
	formattingReady: boolean;
};

/**
 * Extension API event names.
 */
export enum ApiEvent {
	DidRegisterDocumentFormattingEditProvider = 'DidRegisterDocumentFormattingEditProvider',
}
