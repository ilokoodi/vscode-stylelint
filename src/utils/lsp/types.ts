import type LSP from 'vscode-languageserver-protocol';

/**
 * Code actions for a particular rule.
 */
export interface RuleCodeActions {
	/**
	 * Action to disable the rule for a specific line.
	 */
	disableLine?: LSP.CodeAction;

	/**
	 * Action to disable the rule for an entire file.
	 */
	disableFile?: LSP.CodeAction;

	/**
	 * Action to show documentation for the rule.
	 */
	documentation?: LSP.CodeAction;
}
