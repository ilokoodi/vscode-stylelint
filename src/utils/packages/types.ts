import { Stylelint } from '../stylelint/index';

/**
 * Package manager identifiers.
 */
export type PackageManager = 'npm' | 'yarn' | 'pnpm';

/**
 * Options for resolving the Stylelint package.
 */
export type ResolverOptions = {
	packageManager?: PackageManager;
	stylelintPath?: string;
};

/**
 * Stylelint package resolution result.
 */
export type StylelintResolutionResult = {
	stylelint: Stylelint;
	resolvedPath: string;
};

/**
 * A tracer function that can be used to log messages.
 */
export type TracerFn = (message: string, verbose?: string) => void;
