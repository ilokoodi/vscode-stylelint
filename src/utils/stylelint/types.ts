import type LSP from 'vscode-languageserver-protocol';
// eslint-disable-next-line node/no-unpublished-import
import type stylelint from 'stylelint';
import type { PackageManager } from '../packages';

/**
 * Diagnostics for a lint run.
 */
export type LintDiagnostics = {
	/**
	 * The diagnostics, each corresponding to a warning or error emitted by
	 * Stylelint.
	 */
	diagnostics: LSP.Diagnostic[];

	/**
	 * Raw output from Stylelint, if any.
	 */
	output?: string;
};

/**
 * Disable report rule names.
 */
export enum DisableReportRuleNames {
	Needless = '--report-needless-disables',
	InvalidScope = '--report-invalid-scope-disables',
	Descriptionless = '--report-descriptionless-disables',
	Illegal = 'reportDisables',
}

/**
 * Stylelint runner options.
 */
export type RunnerOptions = {
	config?: stylelint.Config | null;
	configBasedir?: string;
	configFile?: string;
	customSyntax?: string;
	ignoreDisables?: boolean;
	packageManager?: PackageManager;
	reportInvalidScopeDisables?: boolean;
	reportNeedlessDisables?: boolean;
	snippet?: string[];
	stylelintPath?: string;
	validate?: string[];
};

/**
 * Error thrown when a rule's option is invalid.
 */
export class InvalidOptionError extends Error {
	reasons: string[];

	constructor(warnings: { text: string }[]) {
		const reasons = warnings.map((warning) => warning.text);

		super(reasons.join('\n'));
		this.reasons = reasons;
	}
}

// TODO: Create type upstream
// https://github.com/stylelint/stylelint/pull/5696
/**
 * A Stylelint configuration error. Taken from
 * https://github.com/stylelint/stylelint/blob/551dcb5/lib/utils/configurationError.js
 */
export type ConfigurationError = Error & { code: 78 };
