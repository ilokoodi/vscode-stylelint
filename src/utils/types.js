'use strict';

const { CodeActionKind: VSCodeActionKind } = require('vscode-languageserver-types');

/**
 * Command IDs
 * @enum {string}
 */
const CommandId = {
	ApplyAutoFix: 'stylelint.applyAutoFix',
};

/**
 * Code action kinds
 * @enum {string}
 */
const CodeActionKind = {
	StylelintSourceFixAll: `${VSCodeActionKind.SourceFixAll}.stylelint`,
};

/**
 * Disable report rule names
 * @enum {string}
 */
const DisableReportRuleNames = {
	Needless: '--report-needless-disables',
	InvalidScope: '--report-invalid-scope-disables',
	Descriptionless: '--report-descriptionless-disables',
	Illegal: 'reportDisables',
};

/**
 * Error thrown when a rule's option is invalid.
 */
class InvalidOptionError extends Error {
	/** @param {{text: string}[]} warnings */
	constructor(warnings) {
		const reasons = warnings.map((warning) => warning.text);

		super(reasons.join('\n'));
		this.reasons = reasons;
	}
}

module.exports = {
	CommandId,
	CodeActionKind,
	DisableReportRuleNames,
	InvalidOptionError,
};
