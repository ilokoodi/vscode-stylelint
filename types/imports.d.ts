import * as postcss from 'postcss';
export import postcss = postcss;

import * as stylelint from 'stylelint';
export import stylelint = stylelint;

import * as vscode from 'vscode';
export import vscode = vscode;

import * as vscodeLanguageServer from 'vscode-languageserver';
export import vscodeLanguageServer = vscodeLanguageServer;

import * as vscodeLanguageServerTextDocument from 'vscode-languageserver-textdocument';
export import vscodeLanguageServerTextDocument = vscodeLanguageServerTextDocument;

import * as vscodeLanguageServerTypes from 'vscode-languageserver-types';
export import vscodeLanguageServerTypes = vscodeLanguageServerTypes;

export as namespace imports;
