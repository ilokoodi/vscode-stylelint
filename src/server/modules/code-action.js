'use strict';

const { CodeActionKind, CodeAction, TextDocumentEdit } = require('vscode-languageserver-types');
const { CodeActionKind: StylelintCodeActionKind } = require('../../utils/types');

/**
 * @implements {LanguageServerModule}
 */
class CodeActionModule {
	static id = 'code-action';

	/**
	 * The language server context.
	 * @type {LanguageServerContext}
	 */
	#context;

	/**
	 * The logger to use, if any.
	 * @type {winston.Logger | undefined}
	 */
	#logger;

	/**
	 * @param {LanguageServerModuleConstructorParameters} params
	 */
	constructor({ context, logger }) {
		this.#context = context;
		this.#logger = logger;
	}

	/**
	 * @param {lsp.TextDocument} document
	 * @returns {boolean}
	 */
	#shouldCodeAction(document) {
		return this.#context.options.validate.includes(document.languageId);
	}

	/**
	 * @returns {Partial<lsp.InitializeResult>}
	 */
	onInitialize() {
		return {
			capabilities: {
				codeActionProvider: {
					codeActionKinds: [CodeActionKind.QuickFix, StylelintCodeActionKind.StylelintSourceFixAll],
				},
			},
		};
	}

	/**
	 * @returns {void}
	 */
	onDidRegisterHandlers() {
		this.#logger?.debug('Registering onCodeAction handler');

		this.#context.connection.onCodeAction(async ({ context, textDocument }) => {
			this.#logger?.debug('Received onCodeAction', { context, uri: textDocument.uri });

			const only = context.only !== undefined ? context.only[0] : undefined;
			const isSource = only === CodeActionKind.Source;
			const isSourceFixAll =
				only === StylelintCodeActionKind.StylelintSourceFixAll ||
				only === CodeActionKind.SourceFixAll;

			if (!isSourceFixAll && !isSource) {
				this.#logger?.debug('Unsupported code action kind, ignoring', { kind: only });

				return [];
			}

			const uri = textDocument.uri;
			const document = this.#context.documents.get(uri);

			if (!document || !this.#shouldCodeAction(document)) {
				if (this.#logger?.isDebugEnabled()) {
					if (!document) {
						this.#logger.debug('Unknown document, ignoring', { uri });
					} else {
						this.#logger.debug('Document should not be validated, ignoring', {
							uri,
							language: document.languageId,
						});
					}
				}

				return [];
			}

			const identifier = { uri: document.uri, version: document.version };
			const edits = await this.#context.getFixes(document);

			const actions = [
				CodeAction.create(
					`Fix all Stylelint auto-fixable problems`,
					{ documentChanges: [TextDocumentEdit.create(identifier, edits)] },
					StylelintCodeActionKind.StylelintSourceFixAll,
				),
			];

			this.#logger?.debug('Returning code actions', { actions });

			return actions;
		});

		this.#logger?.debug('onCodeAction handler registered');
	}
}

module.exports = {
	CodeActionModule,
};
