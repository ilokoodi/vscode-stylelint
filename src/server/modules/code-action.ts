import { CodeActionKind, CodeAction, TextDocumentEdit } from 'vscode-languageserver-types';
import type { TextDocument } from 'vscode-languageserver-textdocument';
import type LSP from 'vscode-languageserver-protocol';
import type winston from 'winston';
import { CodeActionKind as StylelintCodeActionKind } from '../types';
import type {
	LanguageServerContext,
	LanguageServerModuleConstructorParameters,
	LanguageServerModule,
} from '../types';

export class CodeActionModule implements LanguageServerModule {
	static id = 'code-action';

	/**
	 * The language server context.
	 */
	#context: LanguageServerContext;

	/**
	 * The logger to use, if any.
	 */
	#logger: winston.Logger | undefined;

	constructor({ context, logger }: LanguageServerModuleConstructorParameters) {
		this.#context = context;
		this.#logger = logger;
	}

	#shouldCodeAction(document: TextDocument): boolean {
		return this.#context.options.validate.includes(document.languageId);
	}

	onInitialize(): Partial<LSP.InitializeResult> {
		return {
			capabilities: {
				codeActionProvider: {
					codeActionKinds: [CodeActionKind.QuickFix, StylelintCodeActionKind.StylelintSourceFixAll],
				},
			},
		};
	}

	onDidRegisterHandlers(): void {
		this.#logger?.debug('Registering onCodeAction handler');

		this.#context.connection.onCodeAction(async ({ context, textDocument }) => {
			this.#logger?.debug('Received onCodeAction', { context, uri: textDocument.uri });

			const only = context.only !== undefined ? context.only[0] : undefined;
			const isSource = only === CodeActionKind.Source;
			const isSourceFixAll =
				only === StylelintCodeActionKind.StylelintSourceFixAll ||
				only === CodeActionKind.SourceFixAll;

			if (!isSourceFixAll && !isSource) {
				this.#logger?.debug('Unsupported code action kind, ignoring', { kind: only });

				return [];
			}

			const uri = textDocument.uri;
			const document = this.#context.documents.get(uri);

			if (!document || !this.#shouldCodeAction(document)) {
				if (this.#logger?.isDebugEnabled()) {
					if (!document) {
						this.#logger.debug('Unknown document, ignoring', { uri });
					} else {
						this.#logger.debug('Document should not be validated, ignoring', {
							uri,
							language: document.languageId,
						});
					}
				}

				return [];
			}

			const identifier = { uri: document.uri, version: document.version };
			const edits = await this.#context.getFixes(document);

			const actions = [
				CodeAction.create(
					`Fix all Stylelint auto-fixable problems`,
					{ documentChanges: [TextDocumentEdit.create(identifier, edits)] },
					StylelintCodeActionKind.StylelintSourceFixAll,
				),
			];

			this.#logger?.debug('Returning code actions', { actions });

			return actions;
		});

		this.#logger?.debug('onCodeAction handler registered');
	}
}
