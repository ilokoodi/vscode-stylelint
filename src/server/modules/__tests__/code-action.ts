import { CodeActionKind, Position, TextEdit } from 'vscode-languageserver-types';
import type winston from 'winston';
import { CodeActionKind as StylelintCodeActionKind } from '../../types';
import type { LanguageServerModuleConstructorParameters } from '../../types';

import { CodeActionModule } from '../code-action';

const mockContext = {
	connection: { onCodeAction: jest.fn() },
	documents: { get: jest.fn() },
	options: { validate: [] as string[] },
	getFixes: jest.fn(),
};

const mockLogger = {
	debug: jest.fn(),
	isDebugEnabled: jest.fn(() => true),
} as unknown as jest.Mocked<winston.Logger>;

const getParams = (passLogger = false) =>
	({
		context: mockContext,
		logger: passLogger ? mockLogger : undefined,
	} as unknown as LanguageServerModuleConstructorParameters);

describe('CodeActionModule', () => {
	beforeEach(() => {
		mockContext.options.validate = [];
		jest.clearAllMocks();
	});

	test('should be constructable', () => {
		expect(() => new CodeActionModule(getParams())).not.toThrow();
	});

	test('onInitialize should return results', () => {
		const module = new CodeActionModule(getParams());

		expect(module.onInitialize()).toMatchSnapshot();
	});

	test('onDidRegisterHandlers should register a code action handler', () => {
		const module = new CodeActionModule(getParams());

		module.onDidRegisterHandlers();

		expect(mockContext.connection.onCodeAction).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.onCodeAction).toHaveBeenCalledWith(expect.any(Function));
	});

	test('with action kind CodeActionKind.Source, should create code actions', async () => {
		mockContext.documents.get.mockReturnValue({
			uri: 'foo',
			languageId: 'bar',
		});
		mockContext.options.validate = ['bar'];
		mockContext.getFixes.mockReturnValue([TextEdit.insert(Position.create(0, 0), 'text')]);

		const module = new CodeActionModule(getParams(true));

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCodeAction.mock.calls[0][0];

		const result = await handler({
			context: { only: [CodeActionKind.Source] },
			textDocument: { uri: 'foo' },
		});

		expect(result).toMatchSnapshot();
		expect(mockContext.getFixes).toHaveBeenCalledWith({
			languageId: 'bar',
			uri: 'foo',
		});
	});

	test('with action kind StylelintCodeActionKind.StylelintSourceFixAll, should create code actions', async () => {
		mockContext.documents.get.mockReturnValue({
			uri: 'foo',
			languageId: 'bar',
		});
		mockContext.options.validate = ['bar'];
		mockContext.getFixes.mockReturnValue([TextEdit.insert(Position.create(0, 0), 'text')]);

		const module = new CodeActionModule(getParams(true));

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCodeAction.mock.calls[0][0];

		const result = await handler({
			context: { only: [StylelintCodeActionKind.StylelintSourceFixAll] },
			textDocument: { uri: 'foo' },
		});

		expect(result).toMatchSnapshot();
		expect(mockContext.getFixes).toHaveBeenCalledWith({
			languageId: 'bar',
			uri: 'foo',
		});
	});

	test('with action kind CodeActionKind.SourceFixAll, should create code actions', async () => {
		mockContext.documents.get.mockReturnValue({
			uri: 'foo',
			languageId: 'bar',
		});
		mockContext.options.validate = ['bar'];
		mockContext.getFixes.mockReturnValue([TextEdit.insert(Position.create(0, 0), 'text')]);

		const module = new CodeActionModule(getParams(true));

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCodeAction.mock.calls[0][0];

		const result = await handler({
			context: { only: [CodeActionKind.SourceFixAll] },
			textDocument: { uri: 'foo' },
		});

		expect(result).toMatchSnapshot();
		expect(mockContext.getFixes).toHaveBeenCalledWith({
			languageId: 'bar',
			uri: 'foo',
		});
	});

	test('with no action kind, should not attempt to create actions', async () => {
		const module = new CodeActionModule(getParams(true));

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCodeAction.mock.calls[0][0];

		const result = await handler({
			context: {},
			textDocument: { uri: 'foo' },
		});

		expect(result).toStrictEqual([]);
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Unsupported code action kind, ignoring', {
			kind: undefined,
		});
	});

	test('with incorrect action kind, should not attempt to create actions', async () => {
		const module = new CodeActionModule(getParams(true));

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCodeAction.mock.calls[0][0];

		const result = await handler({
			context: {
				only: ['foo'],
			},
			textDocument: { uri: 'foo' },
		});

		expect(result).toStrictEqual([]);
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Unsupported code action kind, ignoring', {
			kind: 'foo',
		});
	});

	test('if no matching document exists, should not attempt to create actions', async () => {
		mockContext.documents.get.mockReturnValue(undefined);

		const module = new CodeActionModule(getParams(true));

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCodeAction.mock.calls[0][0];

		const result = await handler({
			context: { only: [CodeActionKind.Source] },
			textDocument: { uri: 'foo' },
		});

		expect(result).toStrictEqual([]);
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Unknown document, ignoring', { uri: 'foo' });
	});

	test('if document language ID is not in options, should not attempt to create actions', async () => {
		mockContext.documents.get.mockReturnValue({
			uri: 'foo',
			languageId: 'bar',
		});
		mockContext.options.validate = ['baz'];

		const module = new CodeActionModule(getParams(true));

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCodeAction.mock.calls[0][0];

		const result = await handler({
			context: { only: [CodeActionKind.Source] },
			textDocument: { uri: 'foo' },
		});

		expect(result).toStrictEqual([]);
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith(
			'Document should not be validated, ignoring',
			{
				uri: 'foo',
				language: 'bar',
			},
		);
	});

	test('with no debug log level and no valid document, should not attempt to log reason', async () => {
		mockContext.documents.get.mockReturnValue({
			uri: 'foo',
			languageId: 'bar',
		});
		mockContext.options.validate = ['baz'];
		mockLogger.isDebugEnabled.mockReturnValue(false);

		const module = new CodeActionModule(getParams(true));

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCodeAction.mock.calls[0][0];

		const result = await handler({
			context: { only: [CodeActionKind.Source] },
			textDocument: { uri: 'foo' },
		});

		expect(result).toStrictEqual([]);
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Received onCodeAction', {
			context: { only: [CodeActionKind.Source] },
			uri: 'foo',
		});
	});
});
