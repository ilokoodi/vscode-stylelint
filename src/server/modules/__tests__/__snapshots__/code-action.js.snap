// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeActionModule onInitialize should return results 1`] = `
Object {
  "capabilities": Object {
    "codeActionProvider": Object {
      "codeActionKinds": Array [
        "quickfix",
        "source.fixAll.stylelint",
      ],
    },
  },
}
`;

exports[`CodeActionModule with action kind CodeActionKind.Source, should create code actions 1`] = `
Array [
  Object {
    "edit": Object {
      "documentChanges": Array [
        Object {
          "edits": Array [
            Object {
              "newText": "text",
              "range": Object {
                "end": Object {
                  "character": 0,
                  "line": 0,
                },
                "start": Object {
                  "character": 0,
                  "line": 0,
                },
              },
            },
          ],
          "textDocument": Object {
            "uri": "foo",
            "version": undefined,
          },
        },
      ],
    },
    "kind": "source.fixAll.stylelint",
    "title": "Fix all Stylelint auto-fixable problems",
  },
]
`;

exports[`CodeActionModule with action kind CodeActionKind.SourceFixAll, should create code actions 1`] = `
Array [
  Object {
    "edit": Object {
      "documentChanges": Array [
        Object {
          "edits": Array [
            Object {
              "newText": "text",
              "range": Object {
                "end": Object {
                  "character": 0,
                  "line": 0,
                },
                "start": Object {
                  "character": 0,
                  "line": 0,
                },
              },
            },
          ],
          "textDocument": Object {
            "uri": "foo",
            "version": undefined,
          },
        },
      ],
    },
    "kind": "source.fixAll.stylelint",
    "title": "Fix all Stylelint auto-fixable problems",
  },
]
`;

exports[`CodeActionModule with action kind StylelintCodeActionKind.StylelintSourceFixAll, should create code actions 1`] = `
Array [
  Object {
    "edit": Object {
      "documentChanges": Array [
        Object {
          "edits": Array [
            Object {
              "newText": "text",
              "range": Object {
                "end": Object {
                  "character": 0,
                  "line": 0,
                },
                "start": Object {
                  "character": 0,
                  "line": 0,
                },
              },
            },
          ],
          "textDocument": Object {
            "uri": "foo",
            "version": undefined,
          },
        },
      ],
    },
    "kind": "source.fixAll.stylelint",
    "title": "Fix all Stylelint auto-fixable problems",
  },
]
`;
